# 真核生物基因组的基因预测和注释

## 1 用maker进行真核基因组的基因预测

> maker结合了重复序列屏蔽，augustus，genemark，snap从头预测，同源注释以及RNA-seq的方法，其本身并不属于一种预测基因的方法，其为一套整合各个方法的流程。

### 1.1 获取maker的配置文件

```r
maker -CTL
```

### 配置文件

- maker_bopts.ctl 可不用修改

- maker_exe.ctl 执行路径

- maker_opts.ctl 重点修改对象

  

### 1.2 准备maker所需要的预测基因数据库

  

- genome：fasta格式，由上一步组装获得
- EST：EST 原指表达序列标签(expressedsequencetags, *EST*s)，是指从不同组织来源的cDNA序列，在当下可以用组装的转录本代替。
- altest：同源物种的EST，fasta格式
- protein：同源物种的蛋白序列，fasta格式

组装的转录本可以由[Trinity](https://github.com/trinityrnaseq/trinityrnaseq/wiki)获得。

Trinity是[Broad Institute](https://links.jianshu.com/go?to=http%3A%2F%2Fwww.broadinstitute.org%2F)和[Hebrew University of Jerusalem](https://links.jianshu.com/go?to=http%3A%2F%2Fwww.cs.huji.ac.il%2F)开发的RNA-Seq数据 转录组组装工具，包括三个模块，

- Inchworn(尺蠖): 将RNA-seq数据组装成单个转录本，通常是主要转录亚型的全长转录本
- Chrysalis(蛹): 这一步将上一步得到contig进行聚类，对于每个聚类构建完整的德布罗意图(*de Bruijin*graph)。每个转录本表示的是给定基因或者一组有着共同序列的基因的全部转录组成。 之后会根据图中不相交的点对全部短读数据进行拆分
- Butterfly(蝴蝶): 并行处理各个图(graph), 追踪每个图中的短读和配对短读的路径，最后报告可变剪切亚型的全长转录本，并且区分出旁系同源基因的转录本

![Trinity](https://upload-images.jianshu.io/upload_images/2013053-e88fc3f8d879519f.png)

其输入结果和输出结果都十分简单。

```bash
Trinity --seqType fq --max_memory 50G  \
         --left condA_1.fq.gz,condB_1.fq.gz,condC_1.fq.gz \
         --right condA_2.fq.gz,condB_2.fq.gz,condC_2.fq.gz \
         --CPU 6  --min_kmer_cov 2 \
         --jaccard_clip
```

为加快速度，可以用--normalize_max_read_cov减少高表达的基因的reads数，用```--min_kmer_cov 2```参数丢弃uniquely occurring kmer, 从而降低内存消耗,也可以用fastp去除低质量的reads，用cd-hit-est来去除重复的reads。

对于真菌转录组数据，由于基因密度较高，需添加--jaccard_clip

最终结果为Trinity.fasta文件

### 1.3 运行maker

```
maker & maker.log &
```
## 2 整合其它软件获得的注释结果

### 2.1 软件配置

1. [RepeatModeler](http://www.repeatmasker.org/RepeatModeler/) and [RepeatMasker](http://www.repeatmasker.org/RMDownload.html) with all dependencies (I used NCBI BLAST) and [RepBase](http://www.girinst.org/repbase/) 
2. MAKER
3. [Augustus](http://bioinf.uni-greifswald.de/augustus/) 
4. [BUSCO](http://busco.ezlab.org/) 
5. [SNAP](http://korflab.ucdavis.edu/software.html) 
6. [BEDtools](https://bedtools.readthedocs.io/en/latest/)

### 2.2 原始数据

1. 组装好的基因组文件 ：genome.fasta
2. 组装好的转录本文件：trinity.fasta
3. 同源的蛋白质序列：protein.fasta(from [NCB](https://www.ncbi.nlm.nih.gov/genome) or [uniport](www.uniprot.org) or [JGI](https://genome.jgi.doe.gov/portal)
4. repeat library 

### 2.3 注释重复序列
1. *De Novo* Repeat Identification

   对于真核生物来说，注释其重复序列对于后期基因组的注释是非常重要的。虽然有很多重复序列的数据库如Repbase，但从头对其基因组进行重复序列注释依然是十分有必要的。

   通过RepeatModeler，单核单线程注释几十兆的小型基因组只需要几个小时。

   ```bash
   BuildDatabase -name test -engine ncbi genome.fasta
   RepeatModeler -pa 1 -engine ncbi -database test 
   ```

   得到两个结果文件：

   - test-families.fa  : Consensus sequence
   - test-families.stk : Seed alignments

2. Full Repeat Annotation

   通过RepeatMasker，可以进行更加深入的重复序列注释。我们需要用到上一步的test-families.fa文件

   ```bash
   RepeatMasker -pa 1 -e ncbi -lib test-families.fa -dir mask01/ genome.fasta
   ```

   可以获得4个文件：

   - test.fasta.tbl
   - test.fasta.cat.gz
   - test.fasta.masked
   - test.fasta.out

   针对这4个文件，我们可以结合Repbase数据库，进行第二轮重复序列注释

   ```bash
   RepeatMasker -pa 1 -e ncbi -species fungi -dir mask02 test.fasta.masked
   ```

   同样可以获得4个文件

   - test.fasta.out.out
   - test.fasta.masked.masked
   - test.fasta.masked.cat.gz
   - test.fasta.masked.tbl

   通过一些脚本，让这些注释文件合并，用于输入给maker
   
   ```bash
   cp test.fasta.masked.masked full_mask/test.full_mask.fa
   cp test.fasta.out.out full_mask/test.full_mask.out
   gunzip test.fasta.cat.gz test.fasta.masked.cat.gz 
   cat *.cat > full_mask/test.full_mask.cat
   cd full_mask
   ProcessRepeats -species fungi test.full_mask.cat
   rmOutToGFF3.pl test.full_mask.out > test.full_mask.gff3 #generate gff file
   grep -v -e "Satellite" -e ")n" -e "-rich" test.full_mask.gff3 > test.full_mask.out.complex.gff3 # isolate complex repeats
   cat test.full_mask.complex.gff3 | perl -ane '$id; if(!/^\#/){@F = split(/\t/, $_); chomp $F[-1];$id++; $F[-1] .= "\;ID=$id"; $_ = join("\t", @F)."\n"} print $_' > test.full_mask.complex.reformat.gff3
   
   ```
   
   最后生成的gff3文件可以直接用于maker的注释。

## 3 调整maker注释参数

初始化maker

### 3.1 第一轮注释

```bash
maker -CTL
cp maker_opt.ctl round1_maker_opt.ctl
vim round1_maker_opt.ctl
#-----Genome (these are always required)
genome=/home/castoelab/Desktop/daren/boa_annotation/test.fasta #genome sequence (fasta file or fasta embeded in GFF3 file)
organism_type=eukaryotic #eukaryotic or prokaryotic. Default is eukaryotic默认是真核

#-----Re-annotation Using MAKER Derived GFF3
maker_gff= #MAKER derived GFF3 file第一轮注释通常没有gff3文件
est_pass=0 #use ESTs in maker_gff: 1 = yes, 0 = no
altest_pass=0 #use alternate organism ESTs in maker_gff: 1 = yes, 0 = no
protein_pass=0 #use protein alignments in maker_gff: 1 = yes, 0 = no
rm_pass=0 #use repeats in maker_gff: 1 = yes, 0 = no
model_pass=0 #use gene models in maker_gff: 1 = yes, 0 = no
pred_pass=0 #use ab-initio predictions in maker_gff: 1 = yes, 0 = no
other_pass=0 #passthrough anyything else in maker_gff: 1 = yes, 0 = no

#-----EST Evidence (for best results provide a file for at least one)
est=/home/castoelab/Desktop/daren/boa_annotation/trinity.fasta #set of ESTs or assembled mRNA-seq in fasta format转录组的组装结果
altest=/home/castoelab/Desktop/daren/boa_annotation/est/test.est.fasta #EST/cDNA sequence file in fasta format from an alternate organism近源物种的转录组组装结果
est_gff= #aligned ESTs or mRNA-seq from an external GFF3 file
altest_gff= #aligned ESTs from a closly relate species in GFF3 format

#-----Protein Homology Evidence (for best results provide a file for at least one)
protein=/home/castoelab/Desktop/daren/boa_annotation/protein_files_other_squamates/test.fasta  #protein sequence file in fasta format (i.e. from mutiple oransisms)
protein_gff=  #aligned protein homology evidence from an external GFF3 file

#-----Repeat Masking (leave values blank to skip repeat masking)
model_org=simple #select a model organism for RepBase masking in RepeatMasker
rmlib= #provide an organism specific repeat library in fasta format for RepeatMasker
repeat_protein=/opt/maker/data/te_proteins.fasta #provide a fasta file of transposable element proteins for RepeatRunner
rm_gff=/home/castoelab/Desktop/daren/boa_annotation/full_mask/test.full_mask.complex.reformat.gff3 #pre-identified repeat elements from an external GFF3 file
prok_rm=0 #forces MAKER to repeatmask prokaryotes (no reason to change this), 1 = yes, 0 = no
softmask=1 #use soft-masking rather than hard-masking in BLAST (i.e. seg and dust filtering)

#-----Gene Prediction
snaphmm= #SNAP HMM file
gmhmm= #GeneMark HMM file
augustus_species= #Augustus gene prediction species model
fgenesh_par_file= #FGENESH parameter file
pred_gff= #ab-initio predictions from an external GFF3 file
model_gff= #annotated gene models from an external GFF3 file (annotation pass-through)
est2genome=1 #infer gene predictions directly from ESTs, 1 = yes, 0 = no
protein2genome=1 #infer predictions from protein homology, 1 = yes, 0 = no
trna=0 #find tRNAs with tRNAscan, 1 = yes, 0 = no
snoscan_rrna= #rRNA file to have Snoscan find snoRNAs
unmask=0 #also run ab-initio prediction programs on unmasked sequence, 1 = yes, 0 = no

#-----Other Annotation Feature Types (features MAKER doesn't recognize)
other_gff= #extra features to pass-through to final MAKER generated GFF3 file

#-----External Application Behavior Options
alt_peptide=C #amino acid used to replace non-standard amino acids in BLAST databases
cpus=1 #max number of cpus to use in BLAST and RepeatMasker (not for MPI, leave 1 when using MPI)

#-----MAKER Behavior Options
max_dna_len=100000 #length for dividing up contigs into chunks (increases/decreases memory usage)
min_contig=1 #skip genome contigs below this length (under 10kb are often useless)

pred_flank=200 #flank for extending evidence clusters sent to gene predictors
pred_stats=0 #report AED and QI statistics for all predictions as well as models
AED_threshold=1 #Maximum Annotation Edit Distance allowed (bound by 0 and 1)
min_protein=0 #require at least this many amino acids in predicted proteins
alt_splice=0 #Take extra steps to try and find alternative splicing, 1 = yes, 0 = no
always_complete=0 #extra steps to force start and stop codons, 1 = yes, 0 = no
map_forward=0 #map names and attributes forward from old GFF3 genes, 1 = yes, 0 = no
keep_preds=0 #Concordance threshold to add unsupported gene prediction (bound by 0 and 1)

split_hit=10000 #length for the splitting of hits (expected max intron size for evidence alignments)
single_exon=0 #consider single exon EST evidence when generating annotations, 1 = yes, 0 = no
single_length=250 #min length required for single exon ESTs if 'single_exon is enabled'
correct_est_fusion=0 #limits use of ESTs in annotation to avoid fusion genes

tries=2 #number of times to try a contig if there is a failure for some reason
clean_try=0 #remove all data from previous run before retrying, 1 = yes, 0 = no
clean_up=0 #removes theVoid directory with individual analysis files, 1 = yes, 0 = no
TMP= #specify a directory other than the system default temporary directory for temporary files
```

运行maker

```bash
maker -base test_rnd1 round1_maker_opts.ctl maker_bopts.ctl maker_exe.ctl
```

整理第一轮注释的结果

```bash
cd test.maker.out
gff3_merge -d test_rnd1_master_datastore_index.log
fasta_merge -d test_rnd1_master_datastore_index.log
gff3_merge -n -s -d test_rnd1_master_datastore_index.log > test_rnd1.all.maker.noseq.gff #选择不输出sequence
```



### 3.2 第二轮注释

第二轮注释可以用SNAP这个软件，它的速度非常快，可以生成基因的hmm模型，需要用到第一轮注释生成的gff文件

```bash
mkdir snap
cd snap
mv ../test_rnd1.all.maker.noseq.gff .
maker2zff test_rnd1.all.maker.noseq.gff
```

可以得到两个基因组的注释文件，分别是

- genome.ann
- genome.dna

参考[SNAP](https://github.com/KorfLab/SNAP)的官方文档，可以对其进行训练

```bash
fathom genome.ann genome.dna -gene-stats > test.gene.stats.txt
fathom genome.ann genome.dna -validate > test.validation.txt
fathom genome.ann genome.dna -categorize 1000 #将序列分类，这里的1000推荐用基因平均长度的一半，根据基因密度适当调整
fathom uni.ann uni.dna -export 1000 -plus# 只输出unigene中正链基因
```





